Fixed LinkedIn URL + Created two new files 🐱💯

boushrabettir opened the pull request:
- Created `opportunity.py` file which contains a class where one of the functions, `list_all_opportunities()` returns a `List[Opportunity]` as well as prints out all the jobs in the db.

- Created `utility.py` file which, has for now, has a `calculate_day_difference()` function. Will be used to place general reusable helper functions.

- Fixed LinkedIn web scraping issue (incorrect link as well as wrong dates being returned).

boushrabettir committed:
Pushing many changes. Please view description.

- Created `opportunity.py` file which returns a `List[Opportunity]` jobs

- Created `utility.py` file which has for now has a `calculate_day_difference()` function. Will be used to place general reusable helper functions

- Fixed Linkedin webscrapping issue with wrong dates being returned.

boushrabettir committed:
Added file to .env

boushrabettir committed:
Pushing in new requirements.txt file.

Look at title.

jjoeldaniel commented on main.py:
```diff
@@ -148,44 +148,27 @@ def request_linkedin_data() -> List[object]:
         "div",
         class_="base-card relative w-full hover:no-underline focus:no-underline base-card--link base-search-card base-search-card--link job-search-card",
```
Is it possible to narrow this down to avoid issues down the line where classes change?

boushrabettir replied:
Sure!

26samaahmed commented:
gym bro made it!

boushrabettir committed:
Pushing in more changes. Check description please.

- Separated opportunity related functions away from class in `opportunity.py`

- Utilized functions in `utility.py` as well as `opportunity.py`

- Tested by calling `reset_processed_status()` and seeing if the functions still worked and it did. Made sure to update those opportunities as well so they don't get spewed out by the discord bot.

26samaahmed reviewed (approved):
Excellent organization skills.

EthanThatOneKid commented on utility.py:
```diff
@@ -0,0 +1,17 @@
+import psycopg2
+import os
+
+def instantiates_db_connection():
```
Nit: `instantiates` implies a condition.
```suggestion
def instantiate_db_connection():
```

EthanThatOneKid commented on opportunity.py:
```diff
@@ -0,0 +1,92 @@
+    # TODO
+
+
+def list_all_opportunities() -> List[object]:  # Return type of opportunity
```
Return a list of the `Opportunity` dataclass.
```suggestion
def list_all_opportunities() -> List[Opportunity]:  # Return type of opportunity
```

EthanThatOneKid commented on opportunity.py:
```diff
@@ -0,0 +1,92 @@
+@dataclass
+class Opportunity:
+    """Class for ..."""
+    # TODO
```
Here is an example `Opportunity` dataclass.
```suggestion
class Opportunity:
    _company: str
    _title: str
    _location: str
    _link: str
    _processed: bool
```

EthanThatOneKid commented on opportunity.py:
```diff
@@ -0,0 +1,92 @@
+
+    return list_of_objects
+
+
+
```
Nit: Extra newlines.

EthanThatOneKid commented on opportunity.py:
```diff
@@ -0,0 +1,92 @@
+
+    return list_of_objects
+
+
+
+        
+def list_filtered_opportunities() -> List[object]:
+    """Returns a List[object] of job data that have a status of _processed = 0"""
+
+    with utils.instantiates_db_connection() as connection:
+        cursor = connection.cursor()
+
+        cursor.execute(f"SELECT * FROM {table_name} WHERE _processed = 0 LIMIT 20")
+        rows = cursor.fetchall()
+
+        list_of_objects = []
+
+        for row in rows:
+            job_objects = {}
+
+            job_objects["_company"] = row[0]
+            job_objects["_title"] = row[1]
+            job_objects["_location"] = row[2]
+            job_objects["_link"] = row[3]
+            job_objects["_processed"] = row[4]
+
+                # Uncomment to view up to 10 jobs that have not been posted
+
+                # _company, _title, _location, _link, _processed = row
+
+                # print("Company:", _company)
+                # print("Title:", _title)
+                # print("Location:", _location)
+                # print("Link:", _link)
+                # print("Processed:", _processed)
+                # print(" ")
+
+            list_of_objects.append(job_objects)
+
+    return list_of_objects
```
Let's use the `Opportunity` dataclass and abstract the reading of rows into a list of opportunities. Here is a starting point.
```suggestion
def list_all_opportunities() -> List[Opportunity]:
    """Lists all opportunities in DB as well as returns them"""
    with utils.instantiates_db_connection() as connection:
        cursor = connection.cursor()
        cursor.execute(f"SELECT * FROM {table_name}")
        rows = cursor.fetchall()
        return read_opportunities(rows)

def list_filtered_opportunities() -> List[Opportunity]:
    """Returns a List[Opportunity] of job data that have a status of _processed = 0"""
    with utils.instantiates_db_connection() as connection:
        cursor = connection.cursor()
        cursor.execute(f"SELECT * FROM {table_name} WHERE _processed = 0 LIMIT 20")
        rows = cursor.fetchall()
        return read_opportunities(rows)

def read_opportunities(rows: List[tuple], debug: bool) -> List[Opportunity]:
    opps = []
    for row in rows: 
        _company, _title, _location, _link, _processed = row
        if debug:
            print("Company:", _company)
            print("Title:", _title)
            print("Location:", _location)
            print("Link:", _link)
            print("Processed:", _processed)
            print(" ")
        opp = Opportunity(_company, _title, _location, _link, _processed)
        opps.append(opp)
    return opps
```

EthanThatOneKid reviewed (commented):
Thank you for continuing progress on this project! For now, I left a few comments about the `Opportunity` dataclass.

boushrabettir committed:
Many changes made. Please view description! :cat:

- Updated instantiates_db_connection() to instantiate_db_connection()

- Moved update_opportunities_status() to opportunity.py file as it was more relevant

- Added three member functions called, gpt_job_analyzer(), parse_gpt_values(), and determine_job_list() where all those functions help in filtering job opportunities using gpt4free. Please view comments in `opportunity.py` for more details. So far, there is a bug within the response since it returns a string, so I am thinking of putting it into a code block inorder to parse it correctly.

- Placed comment in main.py where we would place the function `gpt_job_analyzer()` before ingesting the opportunities

- Need to fix workflow in automate.yaml file

boushrabettir commented:
Also, theres an [`enhancement issue`](https://github.com/acmcsufoss/acmOpportunities/issues/5#issue-1729312772) open that I'm going to reference since it's relevant.

boushrabettir committed:
Several, several changes made. Please view description.

- Moved ingest_opportunities() to `opportunity.py` file

- Updated the calculate_day_difference() to a method that I found out in Python called `.toordinal()` which better calculates day difference. Link is provided.

- The opportunity functions which recieve a List[Opportunity] and traverse this list, selecting the job descriptions were changed from, for instance, data_block["_company"] to data_block._company

- In `opportunity.py` enhanced the prompt in gpt_job_analyzer(). Seems repetitivness in the prompt is the way to go.

- Within parse_gpt_values() I decided to use ast using .literal_eval(), but I didn't know where to go from there, so I asked ChatGPT for better guidance on what to do next, and learned the isinstance() method (returns T/F)

boushrabettir commented:
Theres no big "visual" difference, but with all my changes the message still was able to go through despite all my [changes](https://github.com/acmcsufoss/acmOpportunities/pull/7/commits/6bc7c18f27882352d99bcc58c69217548ce5edc8). This is just testing LinkedIn data. 

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/e35cf9e2-c465-4e4d-a183-48d5ba38efe1)

boushrabettir commented:
Also to show the changes that GPT did, I forgot to send in this (Only used LinkedIn since either or with Indeed or LinkedIn it'd do the same thing):

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/9eff93ea-0cbc-49c6-8f0a-6f4a87407ea0)

boushrabettir committed:
Updates made. Bot still works despite changes.

- Updated the job extractor functions to return a List[Opportunity] rather than a plain object

- Updated functions in opportunity.py so it successfully parses the response from GPT, and it also keeps looping until we receive a valid response

boushrabettir committed:
Mild README.md file update.

EthanThatOneKid commented on .github/workflows/automate.yaml:
```diff
@@ -2,7 +2,8 @@ name: Daily Data Retrieval Automation
 on:
   schedule:
-    - cron: "0 0 */2 * *" # Runs every other day at 12AM
+    # TODO - Fix this time issue
```
Create a new issue as to not forget to follow up on this in the future.
```suggestion
    # TODO(#<ISSUE NUMBER>): Fix this time issue.
```

boushrabettir replied:
Resolved [here](https://github.com/acmcsufoss/acmOpportunities/pull/7/commits/e873aa82ae04037e5f55d8141c2e11e7f4be4012)

EthanThatOneKid commented on opportunity.py:
```diff
@@ -0,0 +1,173 @@
+    return parsed_values
+
+
+def gpt_job_analyzer(list_of_opps) -> List[Opportunity]:
```
Does the introduction of `gpt_job_analyzer` resolve <https://github.com/acmcsufoss/acmOpportunities/issues/5>?

boushrabettir replied:
Yes, it does! I was waiting for the review before I closed it.

EthanThatOneKid commented on utility.py:
Does the introduction of `utility.py` resolve <https://github.com/acmcsufoss/acmOpportunities/issues/8>?

boushrabettir replied:
I mean, it technically could be the case, but `utility.py` could be used for more general functions that could be helper functions versus having a db file reserved only for db helper functions but, we could implement those in `utility.py`.

EthanThatOneKid replied:
Would it help to move the database helpers into a `db.py`?

EthanThatOneKid replied:
If so, we can save that for a future PR and leave #8 open until we create a `db.py`.

EthanThatOneKid reviewed (approved):
The progress on this PR is spectacular 🤯. I noticed it potentially resolves multiple issues, so be sure that all issues that are to be resolved by this PR are listed in the "Development" list. Be sure `main.py` runs smoothly as expected in manual local testing before merging.

boushrabettir commented:
> . I noticed it potentially resolves multiple issues, so be sure that all issues that are to be resolved by this PR are listed in the "Development" list. Be sure `main.py` runs smoothly as expected in manual local testing before merging

There is... I have a photo here which I tested every function! (even ingesting new opportunities in the DB. I did not update their status though since they haven't been used by the ACM Bot.) 

![image](https://github.com/acmcsufoss/acmOpportunities/assets/116927138/78046ef1-654e-4293-a095-b4da8ac8ab16)

boushrabettir committed:
Updated comment with Issue in yaml file.
